File: experiments/evm_feasibility/native_v1_custody_reference.py
#!/usr/bin/env python3
"""Reproduce pinned V1 undelegation custody fixtures; use --record to update."""

import argparse
import hashlib
import json
import os
import pathlib
import subprocess
import tempfile

from reference import REVISIONS, ROOT

HERE = pathlib.Path(__file__).resolve().parent
SOURCE = pathlib.Path(os.environ.get("TARAXA_EVM_SOURCE", ROOT / "submodules/taraxa-evm"))
GO_SOURCE = HERE / "native_v1_custody_reference.go"
SUPPORT_SOURCE = HERE / "native_simulation_reference.go"
FIXTURES = HERE / "fixtures/native_v1_custody"
TRACE_TARGET = pathlib.Path("taraxa/state/state_evm/account.go")
TRACE_SHA256 = "070a30bd1f0d8a34ca4e0037346f181a946944f8e38a25b32ec91e68d7577568"
TRACE_NEEDLE = b"func (self *Account) SetStateRawIrreversibly(key *common.Hash, value []byte) {\n"
TRACE_INSERTION = TRACE_NEEDLE + b"\tobserveNativeV1CustodyRawWrite(self.addr, *key, value)\n"
TRACE_HELPER = b'''package state_evm

import "github.com/Taraxa-project/taraxa-evm/common"

var nativeV1CustodyRawWriteObserver func(common.Address, common.Hash, []byte)

// SetNativeV1CustodyRawWriteObserver installs the archive-only synchronous
// observer used to record calls at the irreversible raw-storage boundary.
func SetNativeV1CustodyRawWriteObserver(observer func(common.Address, common.Hash, []byte)) {
	nativeV1CustodyRawWriteObserver = observer
}

func observeNativeV1CustodyRawWrite(address common.Address, key common.Hash, value []byte) {
	if nativeV1CustodyRawWriteObserver != nil {
		nativeV1CustodyRawWriteObserver(address, key, common.CopyBytes(value))
	}
}
'''


def run_reference(revision: str) -> bytes:
    with tempfile.TemporaryDirectory(prefix="rustaxa-native-v1-custody-") as directory:
        tree = pathlib.Path(directory)
        archive = subprocess.check_output(["git", "-C", str(SOURCE), "archive", revision])
        subprocess.run(["tar", "-x", "-C", directory], input=archive, check=True)
        target = tree / TRACE_TARGET
        source = target.read_bytes()
        if hashlib.sha256(source).hexdigest() != TRACE_SHA256 or source.count(TRACE_NEEDLE) != 1:
            raise RuntimeError("V1 custody raw-write trace target changed")
        target.write_bytes(source.replace(TRACE_NEEDLE, TRACE_INSERTION, 1))
        (target.parent / "native_v1_custody_observer.go").write_bytes(TRACE_HELPER)
        command = tree / "cmd/native_v1_custody_reference"
        command.mkdir(parents=True)
        support = SUPPORT_SOURCE.read_text().replace(
            "func main() {", "func nativeSimulationSupportMain() {", 1
        )
        (command / "native_simulation_support.go").write_text(support)
        (command / GO_SOURCE.name).write_bytes(GO_SOURCE.read_bytes())
        return subprocess.check_output(
            ["go", "run", "-mod=readonly", "./cmd/native_v1_custody_reference"], cwd=tree
        )


def validate(label: str, data: bytes) -> dict:
    document = json.loads(data)
    if document["selectors"] != {
        "confirm_undelegate": "45a02561",
        "undelegate": "4d99dd16",
    }:
        raise RuntimeError(f"{label}: selectors changed")
    if document["action_gas"] != {"confirm_undelegate": 20_000, "undelegate": 60_000}:
        raise RuntimeError(f"{label}: native action gas changed")
    scenarios = {scenario["name"]: scenario for scenario in document["scenarios"]}
    expected_names = {
        "magnolia_ficus_partial_cornus_lock",
        "magnolia_ficus_terminal",
        "magnolia_pre_ficus_partial",
        "pre_magnolia_terminal",
        "cacti_lock_priority",
    }
    if set(scenarios) != expected_names:
        raise RuntimeError(f"{label}: scenario set changed")
    dpos_address = "00000000000000000000000000000000000000fe"
    undelegated_topic = "4d10bd049775c77bd7f255195afba5088028ecb3c7c277d393ccff7934f2f92c"
    confirmed_topic = "f8bef3a6fe3b4c932b5b51c6472a89f171d039f4bacf18cff632208938bf0426"
    delegator_topic = "00" * 31 + "d1"
    validator_topic = "00" * 31 + "31"
    iterable_item_key = "0457651292671ad645d4f03ca2fb3583a9f53078f3d5e4518866a2e05dd787b4"
    iterable_position_key = "1e9c3648dd8b43db795c3f2a7cfd5233b9685286284c97694d9eef8a158718c0"
    iterable_count_key = "0700a38280c043ad1117478c0f798c0a070737b0f630657c201d9494170c65ea"
    validator_key = "4ec8e987ae8c32d5e6a93d7f1a07e4068caa2697c4e26d8444fd89965bf0660a"
    expected_before = {
        "period": 0, "delegator_balance": "1000", "contract_balance": "1000",
        "validator_stake": "1000", "total_delegated": "1000", "validator_exists": True,
    }
    for scenario in scenarios.values():
        transactions = {transaction["name"]: transaction for transaction in scenario["transactions"]}
        nonces = [transaction["nonce"] for transaction in scenario["transactions"]]
        if nonces != ["0", "1", "2", "3", "4"]:
            raise RuntimeError(f"{label}/{scenario['name']}: nonce sequence changed")
        if (
            transactions["undelegate"]["gas_used"] != 81_720
            or transactions["duplicate_v1"]["gas_used"] != 81_656
        ):
            raise RuntimeError(f"{label}/{scenario['name']}: undelegate gas changed")
        confirm_names = ("early_confirm", "mature_confirm", "missing_confirm")
        if any(transactions[name]["gas_used"] != 41_464 for name in confirm_names):
            raise RuntimeError(f"{label}/{scenario['name']}: confirmation gas changed")
        if transactions["undelegate"]["execution_error"] or not transactions["undelegate"]["ordered_raw_writes"]:
            raise RuntimeError(f"{label}/{scenario['name']}: undelegate failed")
        if transactions["duplicate_v1"]["execution_error"] != "Undelegation already exist":
            raise RuntimeError(f"{label}/{scenario['name']}: duplicate error changed")
        if transactions["early_confirm"]["execution_error"] != "Undelegation is not yet ready to be withdrawn":
            raise RuntimeError(f"{label}/{scenario['name']}: early error changed")
        if transactions["mature_confirm"]["execution_error"] or not transactions["mature_confirm"]["ordered_raw_writes"]:
            raise RuntimeError(f"{label}/{scenario['name']}: mature confirmation failed")
        if transactions["missing_confirm"]["execution_error"] != "Undelegation does not exist":
            raise RuntimeError(f"{label}/{scenario['name']}: missing error changed")
        for failure in ("duplicate_v1", "early_confirm", "missing_confirm"):
            if transactions[failure]["ordered_raw_writes"] or transactions[failure]["logs"]:
                raise RuntimeError(f"{label}/{scenario['name']}: failed operation emitted effects")
        for transaction in transactions.values():
            if transaction["consensus_error"] or transaction["output"]:
                raise RuntimeError(f"{label}/{scenario['name']}: consensus/output behavior changed")
        if scenario["cacti"] and scenario["unlock_period"] != 8:
            raise RuntimeError(f"{label}: Cacti lock did not take priority")
        if not scenario["cacti"] and scenario["unlock_period"] != 4:
            raise RuntimeError(f"{label}: Cornus lock changed")
        amount = int(scenario["amount"])
        amount_word = amount.to_bytes(32, "big").hex()
        expected_rlp = {
            (300, 4): "c482012c04",
            (1000, 4): "c48203e804",
            (300, 8): "c482012c08",
        }[(amount, scenario["unlock_period"])]
        undelegate_writes = transactions["undelegate"]["ordered_raw_writes"]
        expected_queue_create = [
            {
                "address": dpos_address,
                "key": scenario["v1_object_key"],
                "value": expected_rlp,
            },
            {"address": dpos_address, "key": iterable_item_key, "value": "00" * 19 + "31"},
            {"address": dpos_address, "key": iterable_position_key, "value": "01000000"},
            {"address": dpos_address, "key": iterable_count_key, "value": "01000000"},
        ]
        if undelegate_writes[-4:] != expected_queue_create:
            raise RuntimeError(f"{label}/{scenario['name']}: V1 RLP/key/create order changed")
        expected_queue_remove = [
            {"address": dpos_address, "key": scenario["v1_object_key"], "value": ""},
            {"address": dpos_address, "key": iterable_item_key, "value": ""},
            {"address": dpos_address, "key": iterable_position_key, "value": ""},
            {"address": dpos_address, "key": iterable_count_key, "value": "00000000"},
        ]
        mature_writes = transactions["mature_confirm"]["ordered_raw_writes"]
        if mature_writes[:4] != expected_queue_remove:
            raise RuntimeError(f"{label}/{scenario['name']}: V1 queue removal order changed")
        expected_topics = [delegator_topic, validator_topic]
        successful_events = (
            ("undelegate", undelegated_topic),
            ("mature_confirm", confirmed_topic),
        )
        for name, event_topic in successful_events:
            logs = transactions[name]["logs"]
            if len(logs) != 1 or logs[0] != {
                "address": dpos_address, "topics": [event_topic, *expected_topics], "data": amount_word,
            }:
                raise RuntimeError(f"{label}/{scenario['name']}: {name} log changed")
        if scenario["before"] != expected_before:
            raise RuntimeError(f"{label}/{scenario['name']}: seeded custody state changed")
        remaining = 1000 - amount
        expected_after_undelegate = {
            "period": 1, "delegator_balance": "1000", "contract_balance": "1000",
            "validator_stake": str(remaining), "total_delegated": str(remaining),
            "validator_exists": amount != 1000 or scenario["magnolia"],
        }
        expected_after_confirm = {
            "period": scenario["unlock_period"], "delegator_balance": str(1000 + amount),
            "contract_balance": str(1000 - amount), "validator_stake": str(remaining),
            "total_delegated": str(remaining), "validator_exists": amount != 1000,
        }
        if (
            scenario["after_undelegate_block"] != expected_after_undelegate
            or scenario["after_confirm_block"] != expected_after_confirm
        ):
            raise RuntimeError(f"{label}/{scenario['name']}: custody balance/stake transition changed")

    ficus_partial = scenarios["magnolia_ficus_partial_cornus_lock"]
    pre_ficus_partial = scenarios["magnolia_pre_ficus_partial"]
    ficus_transactions = {
        transaction["name"]: transaction for transaction in ficus_partial["transactions"]
    }
    pre_ficus_transactions = {
        transaction["name"]: transaction
        for transaction in pre_ficus_partial["transactions"]
    }
    ficus_mature = ficus_transactions["mature_confirm"]
    pre_ficus_mature = pre_ficus_transactions["mature_confirm"]
    if (
        len(ficus_mature["ordered_raw_writes"]) != 5
        or ficus_mature["ordered_raw_writes"][4]["key"] != validator_key
    ):
        raise RuntimeError(f"{label}: Ficus validator-count rewrite changed")
    if len(pre_ficus_mature["ordered_raw_writes"]) != 4:
        raise RuntimeError(f"{label}: pre-Ficus confirmation unexpectedly rewrote validator count")
    magnolia_terminal = scenarios["magnolia_ficus_terminal"]
    pre_magnolia_terminal = scenarios["pre_magnolia_terminal"]
    magnolia_transactions = {
        transaction["name"]: transaction
        for transaction in magnolia_terminal["transactions"]
    }
    pre_magnolia_transactions = {
        transaction["name"]: transaction
        for transaction in pre_magnolia_terminal["transactions"]
    }
    if (
        len(magnolia_transactions["undelegate"]["ordered_raw_writes"]) != 13
        or len(magnolia_transactions["mature_confirm"]["ordered_raw_writes"]) != 13
    ):
        raise RuntimeError(f"{label}: Magnolia terminal cleanup placement changed")
    if (
        len(pre_magnolia_transactions["undelegate"]["ordered_raw_writes"]) != 19
        or len(pre_magnolia_transactions["mature_confirm"]["ordered_raw_writes"]) != 4
    ):
        raise RuntimeError(f"{label}: pre-Magnolia terminal cleanup placement changed")
    return document


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--record", action="store_true")
    args = parser.parse_args()
    artifacts = {label: run_reference(revision) for label, revision in REVISIONS.items()}
    for label, data in artifacts.items():
        validate(label, data)
    if artifacts["public"] != artifacts["local"]:
        raise RuntimeError("pinned V1 custody references diverged")
    manifest = {
        "schema": 1, "references": REVISIONS,
        "go_version": subprocess.check_output(["go", "version"], text=True).strip(),
        "scope": (
            "Actual StateTransition V1 undelegate/confirm custody, ordered raw-write "
            "instrumentation, Cornus/Cacti locks and Magnolia/Ficus cleanup boundaries; "
            "no production routing claim"
        ),
        "instrumentation": {
            "target": str(TRACE_TARGET),
            "sha256": TRACE_SHA256,
            "observer_sha256": hashlib.sha256(TRACE_HELPER).hexdigest(),
            "insertion": (
                "synchronous copy-only observer before SetStateRawIrreversibly mutation body"
            ),
        },
        "support_sha256": hashlib.sha256(SUPPORT_SOURCE.read_bytes()).hexdigest(),
        "public_local_byte_identical": artifacts["public"] == artifacts["local"],
        "sha256": {label: hashlib.sha256(data).hexdigest() for label, data in artifacts.items()},
        "exporter_sha256": hashlib.sha256(GO_SOURCE.read_bytes()).hexdigest(),
    }
    if args.record:
        FIXTURES.mkdir(parents=True, exist_ok=True)
        for label, data in artifacts.items():
            (FIXTURES / f"{label}.json").write_bytes(data)
        (FIXTURES / "manifest.json").write_text(json.dumps(manifest, indent=2) + "\n")
    else:
        saved = json.loads((FIXTURES / "manifest.json").read_text())
        for field in (
            "schema",
            "references",
            "scope",
            "instrumentation",
            "support_sha256",
            "public_local_byte_identical",
            "sha256",
            "exporter_sha256",
        ):
            if saved[field] != manifest[field]:
                raise RuntimeError(f"V1 custody manifest mismatch: {field}")
        for label, data in artifacts.items():
            if data != (FIXTURES / f"{label}.json").read_bytes():
                raise RuntimeError(f"V1 custody fixture mismatch: {label}")
    print("Both pinned V1 custody references " + ("recorded" if args.record else "reproduced"))
    print("Public/local byte-identical:", manifest["public_local_byte_identical"])


if __name__ == "__main__":
    main()

File: experiments/evm_feasibility/native_v1_custody_reference.go
// Pinned V1 undelegation custody oracle. The harness pairs this exporter with
// the versioned in-memory StateTransition support from native_simulation_reference.go
// and an archive-only observer at the raw-storage write boundary.
package main

import (
	"bytes"
	"encoding/hex"
	"encoding/json"
	"math/big"
	"os"
	"sync"

	"github.com/Taraxa-project/taraxa-evm/common"
	"github.com/Taraxa-project/taraxa-evm/core"
	"github.com/Taraxa-project/taraxa-evm/core/types"
	"github.com/Taraxa-project/taraxa-evm/core/vm"
	"github.com/Taraxa-project/taraxa-evm/params"
	"github.com/Taraxa-project/taraxa-evm/taraxa/state/chain_config"
	dpos "github.com/Taraxa-project/taraxa-evm/taraxa/state/contracts/dpos/precompiled"
	contract_storage "github.com/Taraxa-project/taraxa-evm/taraxa/state/contracts/storage"
	"github.com/Taraxa-project/taraxa-evm/taraxa/state/state_db"
	"github.com/Taraxa-project/taraxa-evm/taraxa/state/state_evm"
	"github.com/Taraxa-project/taraxa-evm/taraxa/state/state_transition"
)

var (
	custodyDelegator = common.HexToAddress("0x00000000000000000000000000000000000000d1")
	custodyOwner     = common.HexToAddress("0x00000000000000000000000000000000000000a1")
	custodyValidator = common.HexToAddress("0x0000000000000000000000000000000000000031")
)

type custodyRawWrite struct {
	Address string `json:"address"`
	Key     string `json:"key"`
	Value   string `json:"value"`
}

var custodyWrites struct {
	sync.Mutex
	values []custodyRawWrite
}

func beginCustodyWrites() {
	custodyWrites.Lock()
	custodyWrites.values = make([]custodyRawWrite, 0)
	custodyWrites.Unlock()
	state_evm.SetNativeV1CustodyRawWriteObserver(func(address common.Address, key common.Hash, value []byte) {
		custodyWrites.Lock()
		custodyWrites.values = append(custodyWrites.values, custodyRawWrite{
			Address: hex.EncodeToString(address[:]), Key: hex.EncodeToString(key[:]), Value: hex.EncodeToString(value),
		})
		custodyWrites.Unlock()
	})
}

func finishCustodyWrites() []custodyRawWrite {
	state_evm.SetNativeV1CustodyRawWriteObserver(nil)
	custodyWrites.Lock()
	defer custodyWrites.Unlock()
	result := append([]custodyRawWrite(nil), custodyWrites.values...)
	custodyWrites.values = nil
	return result
}

type custodyLog struct {
	Address string   `json:"address"`
	Topics  []string `json:"topics"`
	Data    string   `json:"data"`
}

type custodyState struct {
	Period           uint64 `json:"period"`
	DelegatorBalance string `json:"delegator_balance"`
	ContractBalance  string `json:"contract_balance"`
	ValidatorStake   string `json:"validator_stake"`
	TotalDelegated   string `json:"total_delegated"`
	ValidatorExists  bool   `json:"validator_exists"`
}

type custodyTransaction struct {
	Name           string            `json:"name"`
	Selector       string            `json:"selector"`
	Nonce          string            `json:"nonce"`
	GasUsed        uint64            `json:"gas_used"`
	ConsensusError string            `json:"consensus_error"`
	ExecutionError string            `json:"execution_error"`
	Output         string            `json:"output"`
	Logs           []custodyLog      `json:"logs"`
	RawWrites      []custodyRawWrite `json:"ordered_raw_writes"`
}

type custodyScenario struct {
	Name            string               `json:"name"`
	Amount          string               `json:"amount"`
	Magnolia        bool                 `json:"magnolia"`
	Ficus           bool                 `json:"ficus"`
	CornusLock      uint32               `json:"cornus_lock"`
	Cacti           bool                 `json:"cacti"`
	CactiLock       uint32               `json:"cacti_lock"`
	UnlockPeriod    uint64               `json:"unlock_period"`
	ObjectKey       string               `json:"v1_object_key"`
	IterablePrefix  string               `json:"v1_iterable_prefix"`
	Before          custodyState         `json:"before"`
	AfterUndelegate custodyState         `json:"after_undelegate_block"`
	AfterConfirm    custodyState         `json:"after_confirm_block"`
	Transactions    []custodyTransaction `json:"transactions"`
	CommittedRoot   string               `json:"committed_root"`
}

func custodyConfig(magnolia, ficus, cacti bool) chain_config.ChainConfig {
	balances := core.BalanceMap{custodyDelegator: big.NewInt(2_000), custodyOwner: big.NewInt(1_000)}
	maxSupply := big.NewInt(3_000)
	activation := func(enabled bool) types.BlockNum {
		if enabled {
			return 0
		}
		return types.BlockNumberNIL
	}
	return chain_config.ChainConfig{
		EVMChainConfig: params.ChainConfig{ChainId: 666}, GenesisBalances: balances,
		DPOS: chain_config.DPOSConfig{
			EligibilityBalanceThreshold: big.NewInt(100), VoteEligibilityBalanceStep: big.NewInt(10),
			ValidatorMaximumStake: big.NewInt(1_000_000), MinimumDeposit: big.NewInt(1),
			DelegationDelay: 1, DelegationLockingPeriod: 2, BlocksPerYear: 1, YieldPercentage: 0,
			InitialValidators: []chain_config.GenesisValidator{{
				Address: custodyValidator, Owner: custodyOwner, VrfKey: bytes.Repeat([]byte{0x44}, 32),
				Commission: 100, Delegations: core.BalanceMap{custodyDelegator: big.NewInt(1_000)},
			}},
		},
		Hardforks: chain_config.HardforksConfig{
			FixRedelegateBlockNum: 0, FixClaimAllBlockNum: 0,
			RewardsDistributionFrequency: map[uint64]uint32{0: 1},
			MagnoliaHf:                   chain_config.MagnoliaHfConfig{BlockNum: activation(magnolia), JailTime: 1},
			AspenHf: chain_config.AspenHfConfig{
				BlockNumPartOne: 0, BlockNumPartTwo: types.BlockNumberNIL,
				MaxSupply: maxSupply, GeneratedRewards: new(big.Int),
			},
			FicusHf: chain_config.FicusHfConfig{BlockNum: activation(ficus), PillarBlocksInterval: 1_000},
			CornusHf: chain_config.CornusHfConfig{
				BlockNum: 0, DelegationLockingPeriod: 3, DagGasLimit: 1_000_000, PbftGasLimit: 1_000_000,
			},
			SoleiroliaHf: chain_config.SoleiroliaHfConfig{BlockNum: types.BlockNumberNIL},
			CactiHf: chain_config.CactiHfConfig{
				BlockNum: activation(cacti), DelegationLockingPeriod: 7, JailTime: 1,
			},
		},
	}
}

func custodyCommittedState(database *nativeSimulationDB, cfg chain_config.ChainConfig, period uint64) custodyState {
	storageFactory := func(selected types.BlockNum) contract_storage.StorageReader {
		return state_db.ExtendedReader{Reader: database.GetBlockStateReader(selected)}
	}
	reader := new(dpos.API).Init(cfg).NewReader(types.BlockNum(period), storageFactory)
	accountReader := state_db.ExtendedReader{Reader: database.GetBlockStateReader(types.BlockNum(period))}
	balance := func(address common.Address) string {
		value := new(big.Int)
		accountReader.GetRawAccount(&address, func(encoded []byte) {
			value = state_db.DecodeAccountFromTrie(encoded).Balance
		})
		return value.String()
	}
	exists := len(reader.GetVrfKey(&custodyValidator)) != 0
	stake := new(big.Int)
	if exists {
		stake = reader.GetStakingBalance(&custodyValidator)
	}
	dposAddress := *dpos.ContractAddress()
	return custodyState{
		Period: period, DelegatorBalance: balance(custodyDelegator), ContractBalance: balance(dposAddress),
		ValidatorStake: stake.String(), TotalDelegated: reader.TotalAmountDelegated().String(), ValidatorExists: exists,
	}
}

func custodyInput(signature string, amount *big.Int) []byte {
	input := append(nativeSimulationSelector(signature), make([]byte, 12)...)
	input = append(input, custodyValidator[:]...)
	if amount != nil {
		word := make([]byte, 32)
		amount.FillBytes(word)
		input = append(input, word...)
	}
	return input
}

func runCustodyTransaction(transition *state_transition.StateTransition, nonce uint64, name, signature string, amount *big.Int) custodyTransaction {
	dposAddress := *dpos.ContractAddress()
	input := custodyInput(signature, amount)
	beginCustodyWrites()
	result := transition.ExecuteTransaction(&vm.Transaction{
		From: custodyDelegator, To: &dposAddress, Nonce: new(big.Int).SetUint64(nonce),
		GasPrice: new(big.Int), Gas: 200_000, Value: new(big.Int), Input: input,
	})
	writes := finishCustodyWrites()
	logs := make([]custodyLog, len(result.Logs))
	for index, log := range result.Logs {
		topics := make([]string, len(log.Topics))
		for topicIndex, topic := range log.Topics {
			topics[topicIndex] = hex.EncodeToString(topic[:])
		}
		logs[index] = custodyLog{Address: hex.EncodeToString(log.Address[:]), Topics: topics, Data: hex.EncodeToString(log.Data)}
	}
	return custodyTransaction{
		Name: name, Selector: hex.EncodeToString(input[:4]), Nonce: new(big.Int).SetUint64(nonce).String(), GasUsed: result.GasUsed,
		ConsensusError: string(result.ConsensusErr), ExecutionError: string(result.ExecutionErr),
		Output: hex.EncodeToString(result.CodeRetval), Logs: logs, RawWrites: writes,
	}
}

func advanceCustodyBlock(transition *state_transition.StateTransition, period uint64) {
	transition.BeginBlock(&vm.BlockInfo{Author: custodyValidator, GasLimit: 1_000_000, Difficulty: new(big.Int)})
	if uint64(transition.BlockNumber()) != period {
		panic("unexpected custody period")
	}
}

func finishCustodyBlock(transition *state_transition.StateTransition) {
	transition.EndBlock()
	transition.Commit()
}

func runCustodyScenario(name string, amount int64, magnolia, ficus, cacti bool) custodyScenario {
	cfg := custodyConfig(magnolia, ficus, cacti)
	database := newNativeSimulationDB()
	transition := nativeSimulationTransition(database, &cfg)
	defer transition.Close()
	unlock := uint64(4)
	if cacti {
		unlock = 8
	}
	objectKey := contract_storage.Stor_k_1([]byte{3, 0}, custodyValidator[:], custodyDelegator[:])
	iterablePrefix := append([]byte{3, 1}, custodyDelegator[:]...)
	transactions := make([]custodyTransaction, 0, 5)

	advanceCustodyBlock(transition, 1)
	before := custodyCommittedState(database, cfg, 0)
	transactions = append(transactions,
		runCustodyTransaction(transition, 0, "undelegate", "undelegate(address,uint256)", big.NewInt(amount)),
		runCustodyTransaction(transition, 1, "duplicate_v1", "undelegate(address,uint256)", big.NewInt(1)),
		runCustodyTransaction(transition, 2, "early_confirm", "confirmUndelegate(address)", nil),
	)
	finishCustodyBlock(transition)
	afterUndelegate := custodyCommittedState(database, cfg, 1)
	for period := uint64(2); period < unlock; period++ {
		advanceCustodyBlock(transition, period)
		finishCustodyBlock(transition)
	}
	advanceCustodyBlock(transition, unlock)
	transactions = append(transactions,
		runCustodyTransaction(transition, 3, "mature_confirm", "confirmUndelegate(address)", nil),
		runCustodyTransaction(transition, 4, "missing_confirm", "confirmUndelegate(address)", nil),
	)
	finishCustodyBlock(transition)
	afterConfirm := custodyCommittedState(database, cfg, unlock)
	return custodyScenario{
		Name: name, Amount: big.NewInt(amount).String(), Magnolia: magnolia, Ficus: ficus,
		CornusLock: 3, Cacti: cacti, CactiLock: 7, UnlockPeriod: unlock,
		ObjectKey: hex.EncodeToString(objectKey[:]), IterablePrefix: hex.EncodeToString(iterablePrefix),
		Before: before, AfterUndelegate: afterUndelegate, AfterConfirm: afterConfirm, Transactions: transactions,
		CommittedRoot: hex.EncodeToString(database.descriptor.StateRoot[:]),
	}
}

func main() {
	document := map[string]any{
		"schema":     1,
		"selectors":  map[string]string{"undelegate": "4d99dd16", "confirm_undelegate": "45a02561"},
		"action_gas": map[string]uint64{"undelegate": 60_000, "confirm_undelegate": 20_000},
		"scenarios": []custodyScenario{
			runCustodyScenario("magnolia_ficus_partial_cornus_lock", 300, true, true, false),
			runCustodyScenario("magnolia_ficus_terminal", 1_000, true, true, false),
			runCustodyScenario("magnolia_pre_ficus_partial", 300, true, false, false),
			runCustodyScenario("pre_magnolia_terminal", 1_000, false, true, false),
			runCustodyScenario("cacti_lock_priority", 300, true, true, true),
		},
	}
	if err := json.NewEncoder(os.Stdout).Encode(document); err != nil {
		panic(err)
	}
}
